vuu-ui/packages/vuu-popups/src/notifications/NotificationsProvider.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useState, useContext, useCallback, useEffect } from "react";
+import classNames from "classnames"
+import { getUniqueId } from "@finos/vuu-utils";
+
+import "./notifications.css"
+
+// animation times in milliseconds
+const toastDisplayDuration = 6000;
+const horizontalTransitionDuration = 1000;
+const verticalTransitionDuration = 300;
+
+// toast size in pixels
+const toastHeight = 56;
+const toastWidth = 300;
+const toastContainerContentGap = 10;
+const toastContainerLeftPadding = 10;
+// rightPadding is used together with the toastWidth to compute the toast position 
+// at the beginning and at the end of the animation
+const toastContainerRightPadding = 50;
+
+
+const classBase = "vuuToastNotifications";
+
+export enum NotificationLevel {
+  Info = "info",
+  Success = "success",
+  Warning = "warning",
+  Error = "error",
+}
+
+type Notification = {
+  type: NotificationLevel,
+  header: string,
+  body: string,
+  id: string
+}
+
+export const NotificationsContext = React.createContext<{
+  notify: (notification: Omit<Notification, "id">) => void,
+}>({
+  notify: () => "have you forgotten to provide a NotificationProvider?"
+})
+
+export const NotificationsProvider = (props: {
+  children: JSX.Element | JSX.Element[]
+}) => {
+  const [notifications, setNotifications] = useState<Notification[]>([]);
+
+  const notify = useCallback((notification: Omit<Notification, "id">) => {
+    const newNotification = { ...notification, id: getUniqueId() }
+
+    setNotifications(prev => [...prev, newNotification])
+
+    setTimeout(() => {
+      setNotifications(prev => prev.filter(n => n !== newNotification))
+    }, toastDisplayDuration + horizontalTransitionDuration * 2)
+  }, [])
+
+  return (
+    <NotificationsContext.Provider value={{ notify }}>
+      <div
+        className={`${classBase}-toastContainer`}
+        style={{ width: toastWidth + toastContainerRightPadding + toastContainerLeftPadding }}
+      >
+        {
+          notifications.map((notification, i) =>
+            <ToastNotification
+              top={(toastHeight + toastContainerContentGap) * i}
+              notification={notification}
+              key={notification.id}
+            />
+          )
+        }
+      </div>
+      {props.children}
+    </NotificationsContext.Provider>
+  )
+}
+
+export const useNotifications = () => useContext(NotificationsContext);
+
+type ToastNotificationProps = {
+  top: number,
```

Comment:
If this 'top' is responsible for the positioning, I think I'd rather a 'position' prop than a 'top' prop. And I'd love to see it using co-ordinates or something. But, I'm also happy with it as is, so it's up to you.

Reply:
yes, the props top here is used to position the toast vertically (distance from the top of the container). I don't mind calling it "top" here because that is consistent with the naming of the CSS attribute "top" and allows us to just pass top to the style prop like such: style={{top}} which looks nicer than style={{top: yPosition}}. In this case we only want to pass the vertical position down as a prop as the horizontal position is handled internally by the component as state which is why we are not passing a whole "position" coordinates prop